### PR TITLE
feat: update UI elements to use `getSelectedInternalAccount` selector

### PR DESCRIPTION
**Pull request by montelaidev**
## **Description**
Currently, MetaMask faces tight coupling between its UI and the keyring-controller. The UI heavily relies on the keyring-controller's state while also amalgamating information from various sources, such as preferences and balances.

However, this approach presents some challenges. The keyring-controller must be aware of the UI's limitations, like the need for instant account list provision to avoid lag. Moreover, it takes on the responsibility of adding metadata to accounts, such as the associated keyring type, required for displaying account details.

To address these issues, the introduction of the accounts-controller comes into play as a new abstraction layer between the UI and the keyring-controller. This separation of responsibilities allows the keyring-controller to focus on two main tasks:

- Routing requests to the appropriate keyring.
- Persisting the state of the keyrings.

This PR replaces the usage of `getSelectedAddress` with `getSelectedInternalAccount`

Depends on: https://github.com/MetaMask/metamask-extension/pull/21554

## **Related issues**

Resolves https://github.com/MetaMask/accounts-planning/issues/135

## **Manual testing steps**

1. Use a hardware wallet to test all the flows

## **Screenshots/Recordings**

No UI/UX changes

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've clearly explained:
  - [x] What problem this PR is solving.
  - [x] How this problem was solved.
  - [x] How reviewers can test my changes.
- [x] I’ve indicated what issue this PR is linked to: Fixes #???
- [x] I’ve included tests if applicable.
- [x] I’ve documented any added code.
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
- [x] I’ve properly set the pull request status:
  - [ ] In case it's not yet "ready for review", I've set it to "draft".
  - [ ] In case it's "ready for review", I've changed it from "draft" to "non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

**Comment by gantunesr** [comment]
@metamaskbot update-policies

**Review by k-g-j — DISMISSED** [review]

File: ui/components/app/asset-list/asset-list.js [hunk 1]
```diff
@@ -65,7 +65,7 @@ const AssetList = ({ onClickAsset }) => {
   const trackEvent = useContext(MetaMetricsContext);
   const balance = useSelector(getSelectedAccountCachedBalance);
   const balanceIsLoading = !balance;
-  const selectedAddress = useSelector(getSelectedAddress);
+  const { address: selectedAddress } = useSelector(getSelectedAccount);
```
Comment (owencraston) [inline ui/components/app/asset-list/asset-list.js:68]: should this be `getSelectedInternalAccount`?

File: ui/ducks/send/send.js [hunk 2]
```diff
@@ -45,7 +45,8 @@ import {
   getIsMultiLayerFeeNetwork,
   getEnsResolutionByAddress,
   getSelectedAccount,
-  getSelectedAddress,
+  getSelectedInternalAccount,
+  getSelectedInternalAccountWithBalance,
```
Comment (owencraston) [inline ui/ducks/send/send.js:49]: is there no way to derive all the information we need from just one of these methods?

Comment (gantunesr) [reply ui/ducks/send/send.js:49]: Is that only for the benefit of calling on method instead of two in this file?

**Comment by owencraston** [comment]
The send flow also seems to e stuck when sending money to a snap account.
https://github.com/MetaMask/metamask-extension/assets/22918444/d10f2243-4857-48a2-ae5e-a9461bf91f4d

File: ui/ducks/send/send.js (hunk 2 above)
Comment (owencraston) [reply ui/ducks/send/send.js:49]: it will save a re render since these methods are state calls and have side effects. That would be the only reason.

**Comment by k-g-j** [comment]
> The send flow also seems to be stuck when sending money to a snap account. https://github.com/MetaMask/metamask-extension/assets/22918444/d10f2243-4857-48a2-ae5e-a9461bf91f4d

Upon testing these changes with send flow, I can confirm same issues @owencraston  has where it gives a loading yellow spinner perpetually. Also, the list of accounts after clicking send does not reflect the proper snap name (also observed by @owencraston)

Screenshot 2023-12-21 at 12 52 57 PM

File: ui/ducks/send/send.js (hunk 2 above)
Comment (gantunesr) [reply ui/ducks/send/send.js:49]: Following that logic then we should only have one method that exposes the entire state, by the looks of it, if the balance is not necessary then the selector should not provide it. This is also not a component, it's a thunk action creator generator

**Comment by montelaidev** [comment]
@metamaskbot update-policies

**Comment by owencraston** [comment]
Did a quick qa of the branch and the things look pretty good. I noticed that the account names are incorrect in the send flow which was patched by @gantunesr [in this pr last week](https://github.com/MetaMask/metamask-extension/pull/22482/files).
Screenshot 2024-01-16 at 1 04 57 PM
Screenshot 2024-01-16 at 1 00 10 PM

File: ui/components/app/alerts/unconnected-account-alert/unconnected-account-alert.js [hunk 3]
```diff
@@ -43,16 +42,8 @@ const UnconnectedAccountAlert = () => {
...
-  const selectedIdentityWithName = {
-    ...selectedIdentity,
-    name: internalAccounts.find(
-      (internalAccount) =>
-        internalAccount.address === selectedIdentity?.address,
-    )?.metadata.name,
-  };
-  const selectedAddress = useSelector(getSelectedAddress);
+  const account = useSelector(getSelectedInternalAccount);
+  const { address: selectedAddress } = account;
```
Comment (owencraston) [inline ui/components/app/alerts/unconnected-account-alert/unconnected-account-alert.js:46]: I do not think we should remove the logic for the account names as it results in different names in the account list then in the send flow. A better solution would ideally be to export the account name metadata from this selector.

File: ui/components/multichain/app-footer/app-footer.js [hunk 4]
```diff
@@ -55,7 +55,10 @@ export const AppFooter = () => {
   const activeConnections = location.pathname === CONNECTIONS;
   const isUnlocked = useSelector((state) => state.metamask.isUnlocked);
   const isFullScreen = getEnvironmentType() === ENVIRONMENT_TYPE_FULLSCREEN;
-  const selectedAddress = useSelector(getSelectedAddress);
+  const selectedAccount = useSelector(getSelectedInternalAccount);
+
+  // During onboarding there is no selected internal account
+  const selectedAddress = selectedAccount?.address;
```
Comment (k-g-j) [inline ui/components/multichain/app-footer/app-footer.js:61]: Would `const selectedAddress = useSelector(getSelectedAddress)` ever return null? Would a undefined value have an effect on 
```ts
  const connectedSite = connectedSites[selectedAddress]?.find(
    ({ origin }) => origin === currentTabOrigin,
  );
  ```

Comment (montelaidev) [reply ui/components/multichain/app-footer/app-footer.js:61]: It could during onboarding when there aren't any accounts. The connected sites wouldn't be found, but thats ok since we are still setting up the account.

File: ui/components/app/alerts/unconnected-account-alert/unconnected-account-alert.js (hunk 3 above)
Comment (owencraston) [reply ui/components/app/alerts/unconnected-account-alert/unconnected-account-alert.js:46]: retested with the later commits and this is no longer relevant.

**Review by owencraston — DISMISSED** [review]
tested with an account snap with the send flow and sign data. All the flows seem to work. The selector code changes are nice and minimal and I appreciate the extra tests. 🐐

File: ui/components/multichain/import-tokens-modal/import-tokens-modal.js [hunk 5]
```diff
@@ -459,7 +459,9 @@ export const ImportTokensModal = ({ onClose }) => {
         setShowSymbolAndDecimals(false);
         break;
 
-      case Boolean(identities[standardAddress]):
+      case Boolean(
+        accounts.find((internalAccount) => internalAccount.address === address),
+      ):
```
Comment (k-g-j) [inline ui/components/multichain/import-tokens-modal/import-tokens-modal.js:464]: Why the change here to use the original address and not the standardized version?

**Comment by montelaidev** [comment]
@metamaskbot update-policies

**Review by owencraston — DISMISSED** [review]

**Review by k-g-j — DISMISSED** [review]
Reviewed code and satisfied with the thorough unit tests and resolutions to my initial comments. I tested using both my regular EOA and a snap account for sending and signing flows and did not observe any bugs.

**Comment by montelaidev** [comment]
@SocketSecurity ignore npm/@metamask/keyring-api@3.0.0
@SocketSecurity ignore npm/@metamask/eth-keyring-controller@17.0.1
@SocketSecurity ignore npm/@metamask/keyring-controller@12.2.0
@SocketSecurity ignore npm/@metamask/accounts-controller@10.0.0
@SocketSecurity ignore npm/@metamask/obs-store@9.0.0

Ok bumps in controllers and keyring api

**Comment by montelaidev** [comment]
@SocketSecurity ignore npm/rfdc@1.3.1

The dependency has been verified and confirmed with @FrederikBolding.

**Review by owencraston — APPROVED** [review]

**Review by k-g-j — APPROVED** [review]
